File: 189. Rotate Array/rotate.py
from typing import List

class Solution:
  def reversArray(self, arr: List[int], left: int, right: int):
    while left < right:
      arr[left], arr[right] = arr[right], arr[left]
      left += 1
      right -= 1

  def rotate(self, nums: List[int], k: int) -> None:
    """
    Do not return anything, modify nums in-place instead.
    """
    k = k % len(nums)
    left, right = 0, len(nums) - 1
    self.reversArray(nums, left, right)
    
    left, right = 0, k - 1
    self.reversArray(nums, left, right)
    
    left, right = k, len(nums) - 1
    self.reversArray(nums, left, right)
    
nums = [1, 2, 3, 4, 5, 6, 7]
k = 3
solution = Solution()
print('rotate: ', solution.rotate(nums, k), nums)

File: 189. Rotate Array/rotate.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */
const rotate = (nums, k) => {
  const reverseArray = (arr, left, right) => {
    while (left < right) {
      [arr[left], arr[right]] = [arr[right], arr[left]];
      left++;
      right--;
    }
  }

  k = k % nums.length;
  let left = 0;
  let right = nums.length - 1;
  reverseArray(nums, left, right);

  left = 0;
  right = k - 1;
  reverseArray(nums, left, right);

  left = k;
  right = nums.length - 1;
  reverseArray(nums, left, right);
};

const nums = [1, 2, 3, 4, 5, 6, 7];
const k = 3;
console.log('rotate: ', rotate(nums, k), nums);
